### PR TITLE
feat(stubs): Allow to stub directives (fixes #1800)

### LINKED ISSUE #1800: Bug: unable to mock component directives
Consider following code, typical for VTU v1:

```js
  const createComponent = (propsData = { items }) => {
    wrapper = mount(Breadcrumb, {
      directives: { GlTooltip: createMockDirective('gl-tooltip') },
    });
```

In VTU v1 `v-gl-tooltip` will be mocked with our configuration.

However we can't achieve that with VTU v2 ATM. Usage of `global.directives` will not work in same way - it will register directive at app level, and in Vue3 lookup (obviously) have priority at instance level.

After quick look I think we could fix that by messing with directives in `transformVNodeArgs`.

However, if we will look into `global.directives` inside `tranformVNodeArgs` this might create confusing behavior when `global.directives` behave not like `app.directives` being the only difference among `global` options.

So the question is do we want to treat this like a bug, and hook `global.directives` to mocking directive behavior or more like a missing feature and add new option somewhere in mount config for mocking?

WDYT @cexbrayat @lmiller1990?

### PR DESCRIPTION
**Pull request by xanf**
We're adding new vnode transformer (related to directives) and re-use our architecture for searching stubs to make sure that we still could find elements even after stubbing directives on them

### CODE REVIEW

File: tests/mountingOptions/global.stubs.spec.ts
Comment (cexbrayat) [inline tests/mountingOptions/global.stubs.spec.ts:937]: naive question: shouldn't that use a "real" `script setup` SFC like other tests do? That would seem more bulletproof and close to reality.

File: src/stubs.ts
Comment (xanf) [reply src/stubs.ts:16]: We can't clear it safely, because there could be two `mount`ed wrappers simultaneously and `stubs` is shared among them

However, **Weak**Map give us clearing for free in our case

File: src/utils.ts
Comment (xanf) [reply src/utils.ts:173]: :white_check_mark: thank you. I was adding boolean support later and forgot this one. Hate TypeScript forcing me to do `as` here

File: tests/mountingOptions/global.stubs.spec.ts
Comment (xanf) [reply tests/mountingOptions/global.stubs.spec.ts:899]: We have relevant tests for `false` in components, will add for directive

File: tests/mountingOptions/global.stubs.spec.ts
Comment (xanf) [reply tests/mountingOptions/global.stubs.spec.ts:937]: :white_check_mark: Why not, I have no strong opinion on this one

File: docs/guide/advanced/stubs-shallow-mount.md
Comment (xanf) [reply docs/guide/advanced/stubs-shallow-mount.md:301]: :white_check_mark:

File: docs/guide/advanced/stubs-shallow-mount.md
Comment (xanf) [reply docs/guide/advanced/stubs-shallow-mount.md:303]: @cexbrayat I'm ok with your way, but why do you prefer it? I mean `wrapper.classes()` also work since `h1` is root node

File: src/utils.ts
Comment (xanf) [reply src/utils.ts:173]: :white_check_mark: `Directive | true`

File: src/utils/find.ts
Comment (xanf) [reply src/utils/find.ts:47]: :white_check_mark:

File: src/vnodeTransformers/stubDirectivesTransformer.ts
Comment (xanf) [reply src/vnodeTransformers/stubDirectivesTransformer.ts:31]: :white_check_mark:

File: src/vnodeTransformers/stubDirectivesTransformer.ts
Comment (xanf) [reply src/vnodeTransformers/stubDirectivesTransformer.ts:44]: :white_check_mark:

File: tests/mountingOptions/global.stubs.spec.ts
Comment (xanf) [reply tests/mountingOptions/global.stubs.spec.ts:899]: :white_check_mark:

File: docs/guide/advanced/stubs-shallow-mount.md
Comment (cexbrayat) [reply docs/guide/advanced/stubs-shallow-mount.md:303]: Ha, so why is there an empty span above? I thought the template has several root elements

File: tests/mountingOptions/global.stubs.spec.ts
Comment (xanf) [reply tests/mountingOptions/global.stubs.spec.ts:954]: :bomb: Actually this one demonstrated that directives stubbing won't work with `<script setup>`. I will probably go ahead and create an issue for `<script setup>` compiler (I believe we could easily do that there), but for now I just capture current state of the things

File: src/mount.ts
Comment (xanf) [reply src/mount.ts]: :white_check_mark:

File: src/utils/find.ts
Comment (xanf) [reply src/utils/find.ts]: :white_check_mark:

File: docs/guide/advanced/stubs-shallow-mount.md
Comment (xanf) [reply docs/guide/advanced/stubs-shallow-mount.md:303]: Ouch, my bad with copy-paste! Fixed!

File: docs/guide/advanced/stubs-shallow-mount.md
Comment (xanf) [reply docs/guide/advanced/stubs-shallow-mount.md]: :white_check_mark:

File: src/vnodeTransformers/stubDirectivesTransformer.ts
Comment (cexbrayat) [inline src/vnodeTransformers/stubDirectivesTransformer.ts:19]: rare -> rarely

File: tests/mountingOptions/global.stubs.spec.ts
Comment (cexbrayat) [reply tests/mountingOptions/global.stubs.spec.ts:954]: So this test should be updated to reflect that it does not currently work right?
And ideally, reference the issue you'll create for the compiler, so we can fix it when possible.

File: src/vnodeTransformers/stubDirectivesTransformer.ts
Comment (xanf) [reply src/vnodeTransformers/stubDirectivesTransformer.ts:19]: :white_check_mark:

File: tests/mountingOptions/global.stubs.spec.ts
Comment (xanf) [reply tests/mountingOptions/global.stubs.spec.ts:954]: @cexbrayat nope. `{ setup() ... { ... }}` will still work, `<script setup>` will not. I've added link to https://github.com/vuejs/core/issues/6887 with my proposal to docs to reflect that

File: docs/guide/advanced/stubs-shallow-mount.md
Comment (xanf) [reply docs/guide/advanced/stubs-shallow-mount.md:307]: :white_check_mark:

**Review by cexbrayat — APPROVED** [review]
LGTM

Ideally I'd like another pair of eyes on this before merging it

@freakzlike @lmiller1990 would you mind taking a look when you have some time?

**Review by lmiller1990 — APPROVED** [review]
This all looks good to me!

**Pull request merged**
